### PR TITLE
[TIMOB-20320] Android: Provide generic requestPermission(s) method

### PR DESCRIPTION
**Pull request by stgrosshh**
Add generic permissions request handling for Android Marshmallow
see example usage in tipermissions module
https://github.com/stgrosshh/tipermissions

JIRA: https://jira.appcelerator.org/browse/TIMOB-20320

### CODE REVIEW

**Comment by AppWerft** [comment]
It is very handy to use. Actually we only need requestPermissions(). hasPermission() is only for static test.

Here my test: 

`require('ti.permissions').requestPermissions(['android.permission.RECORD_AUDIO', 'android.permission.WRITE_EXTERNAL_STORAGE'], 0, function(e) {
	if (e.success == !0) startAudioRecorder();
	else alert(L('USER_HAS_DENIED'));
});`

File: android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java [hunk 1]
```diff
@@ -475,6 +492,82 @@ public void onRequestPermissionsResult(int requestCode,
...
+	 * register permission request result callback for activity
+	 *
+	 * @param requestCode request code (8 Bit only) for association of callback with request
+	 * @param callback callback function which receives a KrollDict with success, code, optional message and requestCode
+     * @param context KrollObject as required by async callback pattern
+	 * @param requestedPermissions one or more requested permission
+     */
+	public void registerPermissionRequestCallback(Integer requestCode,
+														 KrollFunction callback,KrollObject context,
+														 String... requestedPermissions) {
```
Comment (ashcoding) [inline android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java]: Line 558 to Line 563, Slight formatting issue.

File: android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java [hunk 2]
```diff
@@ -475,6 +492,82 @@ public void onRequestPermissionsResult(int requestCode,
 		}
 	}
 
+	private void handlePermissionRequestResult(Integer requestCode,String[] permissions, int[] grantResults) {
```
Comment (ashcoding) [inline android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java]: Formatting issue for the params.

**Comment by ashcoding** [comment]
There seems to be slight formatting issues here and there in the code. Could you format the code?
Line 558 to 564. Line 460. Line 495. Line 112. Line 121.
You can refer to this to fix it up http://docs.appcelerator.com/platform/latest/#!/guide/Java_Coding_Standards

Or if you're using an IDE with some format function, that might help speed things up. :)

**Comment by stgrosshh** [comment]
Done formatting as requested.

**Comment by hansemannn** [comment]
@stgrosshh Can you add the docs for your new method? Also, how does this affect current apps? We have to ensure backwards-compatibility by deprecating the old API. Thank you!

**Comment by stgrosshh** [comment]
@hansemannn for documentation pls see the provided JavaDoc. It's meant as a means for module developers primarely. See my tipermissions module https://github.com/stgrosshh/tipermissions for an example of usage.

I deprecated the old permission specific methods in my PR so there would be time for migration. I think some of the Appcelerator provided modules use those right now. Therefore some migration effort would be required though. In the meantime they still work of course.

**Comment by hieupham007** [comment]
Thanks for your contributions! I've refactored the code so we can remove the deprecated internal methods. Please refer to this PR: https://github.com/appcelerator/titanium_mobile/pull/7948

**Pull request closed without merging**